astropy/uncertainty/uncertainty.py

```diff
@@ -0,0 +1,306 @@
+        # the latter helps keep track of slices.
+        self.id = (uncertainty.__array_interface__['data'][0],
+                   uncertainty.__array_interface__['strides'])
+        # Any derived uncertainty will use this dictionary, ensuring there is
+        # a link that keeps the value in memory until it is not used any more.
+        self.derivatives = {self.id: [self, 1.]}
+
+    def __call__(self):
+        """Get the uncertainty.  Returns simply the stored value."""
+        return self.uncertainty
+
+    def __getitem__(self, item):
+        # Particular items are stored in a new instance, with a new ID.
+        return self.__class__(uncertainty=self.uncertainty[item], check=False)
+
+    def __repr__(self):
+        return '{0}({1})'.format(type(self).__name__, self.uncertainty)
+
+
+class DerivedUncertainty:
+    """Derived Uncertainty.
+
+    Parameters
+    ----------
+    derivatives : dict
+        A dictionary keyed on uncertainty IDs, containing both the uncertainty
+        and the derivative with which a variable depends on it.
+
+    Notes
+    -----
+    This class is mostly for internal use by `Variable`.
+    """
+    def __init__(self, derivatives):
+        self.derivatives = derivatives
+
+    @property
+    def uncertainty_components(self):
+        """Iterator over the different components of the uncertainty."""
+        for unc_id, (uncertainty, derivative) in self.derivatives.items():
+            uncertainty = uncertainty.uncertainty
+            component = np.abs(derivative * uncertainty)
+            # The component is generally an array; if any derivatives are
+            # undefined, but the uncertainty is 0, the component is zero too.
+            # [Twice faster is to do: np.isnan(np.max(a))]
+            if np.any(np.isnan(derivative)):
+                component[np.logical_and(np.isnan(derivative),
+                                         uncertainty == 0.)] = 0.
+            yield component
+
+    def __call__(self):
+        """Calculate the uncertainty as the rms of the components."""
+        variance = None
+        for delta in self.uncertainty_components:
+            variance = delta**2 if variance is None else variance + delta**2
+        return np.sqrt(variance)
+
+    def __getitem__(self, item):
+        # Get the item from each of the uncertainties that contribute.
+        # TODO: this needs proper support for slicing, incl. broadcasting, etc.
+        derivatives = {}
+        for unc_id, (uncertainty, derivative) in self.derivatives.items():
+            uncertainty = uncertainty[item]
+            if isiterable(derivative):
+                derivative = derivative[item]
+            derivatives[uncertainty.id] = [uncertainty, derivative]
+        return self.__class__(derivatives)
+
+    def __repr__(self):
+        return '{0}({1})'.format(type(self).__name__, self.derivatives)
+
+
+class Variable(np.ndarray):
```

Comment:
I wonder if there may not be a better name for this class.  "Variable" is a term that is already quite [overloaded](https://en.wikipedia.org/wiki/Variable), and doesn't seem such a good fit for the intended use of this class and could be confusing to the uninitiated. In statistical parlance, the objects represented by (instances of) this class are Random Variates - that is realizations of a Random Variable.  It would also be a good idea to include "Array" in the name to indicate that this is actually a numpy array subclass.  I would therefore suggest `RandomVariateArray` as a potential alternative name.

Reply:
Agreed the name is very poor, but `random` to me suggests there is a random number generator involved, which is of course not the case. Maybe `Measurement`? But then I hope to use the same thing also for derived quantities...  [Wikipedia](https://en.wikipedia.org/wiki/Measurement_uncertainty) suggests `Measurand`, but it is the first I ever saw that term...

Reply:
I'm in favour of `Measurement` or `Variate` or even `VariateArray` over `Variable`

Reply:
My problem with `Variate` remains the association with random number generators - I have not seen it used for measurements or derived quantities. My problem with `Measurement` is that this suggests no dependence on other objects, while the whole point of the class is to include correlations. But `Variable` sounds like an instance in a programme, which is also no good.

What I'll do for now is right on top put an item that we have to decide on the name...